tests/avify4mtest.c

```diff
@@ -0,0 +1,177 @@
+// Copyright 2022 Google LLC. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "apps/shared/y4m.h"
+#include "avif/avif.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static const char kPath[] = "/tmp/avify4mtest.y4m";
+
+//------------------------------------------------------------------------------
+
+// Returns true if image1 and image2 are identical.
+static avifBool compareYUVA(const avifImage * image1, const avifImage * image2)
+{
+    if (image1 == NULL || image2 == NULL || image1->width != image2->width || image1->height != image2->height ||
+        image1->depth != image2->depth || image1->yuvFormat != image2->yuvFormat || image1->yuvRange != image2->yuvRange) {
+        printf("ERROR: input mismatch\n");
+        return AVIF_FALSE;
+    }
+
+    avifPixelFormatInfo formatInfo;
+    avifGetPixelFormatInfo(image1->yuvFormat, &formatInfo);
+    uint32_t uvWidth = (image1->width + formatInfo.chromaShiftX) >> formatInfo.chromaShiftX;
+    uint32_t uvHeight = (image1->height + formatInfo.chromaShiftY) >> formatInfo.chromaShiftY;
+    if (uvWidth < 1)
+        uvWidth = 1;
+    if (uvHeight < 1)
+        uvHeight = 1;
+
+    for (uint32_t plane = 0; plane < (formatInfo.monochrome ? 1 : AVIF_PLANE_COUNT_YUV); ++plane) {
+        const uint32_t pixelByteCount =
+            ((plane == AVIF_CHAN_Y) ? image1->width : uvWidth) * ((image1->depth > 8) ? sizeof(uint16_t) : sizeof(uint8_t));
+        const uint32_t height = (plane == AVIF_CHAN_Y) ? image1->height : uvHeight;
+        for (uint32_t y = 0; y < height; ++y) {
+            if (memcmp(image1->yuvPlanes[plane] + y * image1->yuvRowBytes[plane],
+                       image2->yuvPlanes[plane] + y * image2->yuvRowBytes[plane],
+                       pixelByteCount) != 0) {
+                printf("ERROR: different px at row %" PRIu32 ", channel %" PRIu32 "\n", y, plane);
+                return AVIF_FALSE;
+            }
+        }
+    }
+
+    if (image1->alphaPlane != NULL) {
+        if (image2->alphaPlane == NULL || image1->alphaRange != image2->alphaRange ||
+            image1->alphaPremultiplied != image2->alphaPremultiplied) {
+            printf("ERROR: input mismatch\n");
+            return AVIF_FALSE;
+        }
+        const uint32_t pixelByteCount = image1->width * ((image1->depth > 8) ? sizeof(uint16_t) : sizeof(uint8_t));
+        for (uint32_t y = 0; y < image1->height; ++y) {
+            if (memcmp(image1->alphaPlane + y * image1->alphaRowBytes, image2->alphaPlane + y * image2->alphaRowBytes, pixelByteCount) != 0) {
+                printf("ERROR: different px at row %" PRIu32 ", alpha\n", y);
+                return AVIF_FALSE;
+            }
+        }
+    }
+    return AVIF_TRUE;
+}
+
+//------------------------------------------------------------------------------
+
+// Fills each plane of the image with the maximum allowed value.
```

Comment:
For a full-range pixel, the maximum allowed value is `(1 << image->depth) - 1`, not `255 << (image->depth - 8)`.